src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/ZombieClusterMonitor.scala

```diff
@@ -0,0 +1,116 @@
+    Props(new ZombieClusterMonitor(config, gdDAO, googleProjectDAO, dbRef))
+  }
+
+  sealed trait ZombieClusterMonitorMessage
+  case object DetectZombieClusters extends ZombieClusterMonitorMessage
+}
+
+/**
+  * This monitor periodically sweeps the Leo database and checks for clusters which no longer exist in Google.
+  */
+class ZombieClusterMonitor(config: ZombieClusterConfig, gdDAO: GoogleDataprocDAO, googleProjectDAO: GoogleProjectDAO, dbRef: DbReference) extends Actor with LazyLogging {
+  import context._
+
+  override def preStart(): Unit = {
+    super.preStart()
+    system.scheduler.schedule(config.zombieCheckPeriod, config.zombieCheckPeriod, self, DetectZombieClusters)
+  }
+
+  override def receive: Receive = {
+    case DetectZombieClusters =>
+      // Get active clusters from the Leo DB, grouped by project
+      val zombieClusters = getActiveClustersFromDatabase.flatMap { clusterMap =>
+        clusterMap.toList.flatTraverse { case (project, clusters) =>
+          // Check if the project is active
+          isProjectActiveInGoogle(project).flatMap {
+            case true =>
+              // If the project is active, check each individual cluster
+              logger.debug(s"Project ${project.value} containing ${clusters.size} clusters is active in Google")
+              clusters.toList.traverseFilter { cluster =>
+                isClusterActiveInGoogle(cluster).map {
+                  case true =>
+                    logger.debug(s"Cluster ${project.value} / ${cluster.clusterName.value} is active in Google")
+                    None
+                  case false =>
+                    logger.debug(s"Cluster ${project.value} / ${cluster.clusterName.value} is a zombie!")
+                    Some(cluster)
+                }
+              }
+            case false =>
+              // If the project is inactive, all clusters in the project are zombies
+              logger.debug(s"Project ${project.value} containing ${clusters.size} clusters is inactive in Google")
+              Future.successful(clusters.toList)
+          }
+        }
+      }
+
+      // Error out each detected zombie cluster
+      zombieClusters.flatMap { cs =>
+        logger.info(s"Detected ${cs.size} zombie clusters across ${cs.map(_.googleProject).toSet.size} projects.")
+        cs.traverse { cluster =>
+          handleZombieCluster(cluster)
+        }
+      }
+
+  }
+
+  private def getActiveClustersFromDatabase: Future[Map[GoogleProject, Seq[Cluster]]] = {
+    dbRef.inTransaction {
+      _.clusterQuery.listActive
+    } map { clusters =>
+      clusters.groupBy(_.googleProject)
+    }
+  }
+
+  private def isProjectActiveInGoogle(googleProject: GoogleProject): Future[Boolean] = {
+    // Check the project and its billing info
+    (googleProjectDAO.isProjectActive(googleProject.value) |@| googleProjectDAO.isBillingActive(googleProject.value))
+      .map(_ && _)
+      .recover { case _ => false }
```

Comment:
That's a great point. I definitely want to avoid error-ing clusters on transient google API errors. Maybe it would be safer to NOT retire clusters on API errors, and just log a warning in that case. In other words we only retire clusters we KNOW are missing in Google.

Reply:
Perfect. Should we add a test for this case, too (by mocking the respective DAO to return failed futures for the two calls and make sure we observe the desired behaviour)?